Improve SIP packet detection using heuristic parsing

sorooshm78 opened the pull request:
This PR improves SIP packet detection in PcapPlusPlus by introducing heuristic parsing based on Wireshark’s SIP dissector. SIP messages are now detected from the UDP payload itself, not only when using port 5060.

* Add `static bool dissectSipHeuristic(const uint8_t* data, size_t dataLen)` to detect SIP requests/responses from payload content (Wireshark-style logic)
* Use the new heuristic in `UdpLayer` so SIP packets on non-standard ports are correctly classified
* Preserve existing behavior for non-SIP payloads

Related issue: [https://github.com/seladb/PcapPlusPlus/issues/2022](https://github.com/seladb/PcapPlusPlus/issues/2022)

Note: I’m not yet fully familiar with PcapPlusPlus’ internal structure, so if there are better places, names, or patterns for this logic, I’m happy to adjust the PR based on your feedback

seladb commented on Packet++/header/SipLayer.h:
```diff
@@ -114,6 +117,128 @@ namespace pcpp
+		/// @param[in] data Pointer to the raw data buffer
+		/// @param[in] dataLen Length of the data buffer in bytes
+		/// @return True if the first line matches SIP request/response syntax, false otherwise
+		static bool dissectSipHeuristic(const uint8_t* data, size_t dataLen)
```
We already have `SipRequestFirstLine` and `SipResponseFirstLine` that parse the first line, maybe we could use this instead of adding more logic to parse the first line?

sorooshm78 replied:
I agree that we already have `SipRequestFirstLine` and `SipResponseFirstLine` for parsing the first line, but I think keeping the heuristic logic separate is still necessary because the goals are different.

The `Sip*FirstLine` classes assume we already decided that the payload is SIP, operate on a `Sip*Layer` instance, update internal state (`m_IsComplete`, offsets, logging, etc.), and are meant for full parsing.

In contrast, `dissectSipHeuristic()` is a stateless, side-effect-free check that runs directly on raw data to answer a simpler question: *“does this buffer look like a SIP message at all?”*. This also matches Wireshark’s design, where heuristic detection is separate from the actual SIP dissector that parses the first line and fields.

This separation is particularly important for TCP: when we inspect data per segment, the first line may be incomplete. In that case the heuristic must be able to say “need more data / undecided” without constructing SIP layers or marking anything as invalid, which is a different lifecycle than the existing first-line parsers.

In this pull request I’m not yet handling TCP segmentation or IP fragmentation — the heuristic currently assumes it sees at least one complete first line. I plan to address proper TCP stream reassembly / IP fragmentation handling in a separate follow-up PR.

sorooshm78 commented on Packet++/header/SipLayer.h:
```diff
@@ -114,6 +117,128 @@ namespace pcpp
+		/// @param[in] data Pointer to the raw data buffer
+		/// @param[in] dataLen Length of the data buffer in bytes
+		/// @return True if the first line matches SIP request/response syntax, false otherwise
+		static bool dissectSipHeuristic(const uint8_t* data, size_t dataLen)
```
Sure, I’ll move the implementation to `SipLayer.cpp`.

seladb commented:
@sorooshm78 I just saw we already have most of the logic that you implemented: https://github.com/seladb/PcapPlusPlus/blob/master/Packet%2B%2B/src/UdpLayer.cpp#L113-L120

The 2 things you implemented on top of what we have is parsing the request version and the URI, but are they actually required? 🤔

sorooshm78 commented:
@seladb 
I had a PCAP file where SIP packets with source and destination ports other than 5060 were not being detected, and the parsing process was not being performed. The reason for this was due to the condition in the code: `SipLayer::isSipPort(portDst) || SipLayer::isSipPort(portSrc)`.

Upon reviewing the code in Wireshark, I concluded that further investigation is needed. The logic in Wireshark is essentially designed to search the first line and split it into three tokens. Since the format for request and response is defined in the RFC, this determination was made based on that. Specifically, the RFC defines SIP requests and responses in the following way:

* **SIP Request Line**: `Method SP Request-URI SP SIP-Version CRLF`
* **SIP Response Line**: `SIP-Version SP Status-Code SP Reason-Phrase CRLF`

However, based on your comment, I noticed that part of the logic in your parsing already includes this check, so I decided to build on that.

For SIP responses, everything was straightforward and clear because of the fixed length of the version and the three-digit status code, which made detection easy. I could use the `SipResponseFirstLine::parseStatusCode` and `SipResponseFirstLine::parseVersion` methods because of their static nature.

However, for `SipRequestFirstLine`, it was different. In this class, only the `SipRequestFirstLine::parseMethod` method is static, while methods like `SipRequestFirstLine::parseVersion()` and `SipRequestFirstLine::getUri()` are not static. To use these methods, an instance of the `SipRequestFirstLine` class would need to be created, which, as per the logic, isn't correct because we are still in the phase of identifying the SIP protocol.

Therefore, I added two static methods to the `SipRequestFirstLine` class:

* `SipRequestFirstLine::parseUri(const char* data, size_t dataLen)`
* `SipRequestFirstLine::parseVersion(const char* data, size_t dataLen)`

This was necessary to ensure that the parsing of SIP requests was handled correctly without violating the existing design logic.

Please let me know if you think there's a better way to approach this or if there are any concerns with my changes.

sorooshm78 commented:
@seladb 
The SIP port can be anything, and its default is 5060. I understand your performance concerns, and you’re right that we shouldn’t call the heuristic function for every packet. However, relying only on the port to determine the packet type is not sufficient. Because of that, two approaches come to my mind:

1. We can run the heuristic function only after all port-based detections have been checked. If we can’t identify the packet based on its port, then we call the heuristic functions. This way we still get fast detection when the port is enough, and if we can’t determine the packet type from the port, instead of marking it as “unknown” we run the heuristic functions to identify it. This is exactly what Wireshark does: it first tries fast port-based classification, and if that fails, it calls each protocol’s heuristic dissector to detect the packet more accurately.

2. The second approach that comes to mind, which can be used alongside port-based detection for fast classification, is this: since the SIP request and response formats are well defined, we can read the first 3 characters of the payload. If it’s a response, those 3 characters will always be `SIP`. If it’s a request, the first token is the SIP method, and the set of SIP methods is known. So we can check just the first 3 characters of that token. For example, if the request is `INVITE sip:uri SIP/2.0`, by reading the first 3 characters `INV` we can verify that they match the beginning of a valid SIP method.

* SIP Request Line: `Method SP Request-URI SP SIP-Version CRLF`
* SIP Response Line: `SIP-Version SP Status-Code SP Reason-Phrase CRLF`

seladb commented:
> 1. We can run the heuristic function only after all port-based detections have been checked. If we can’t identify the packet based on its port, then we call the heuristic functions. This way we still get fast detection when the port is enough, and if we can’t determine the packet type from the port, instead of marking it as “unknown” we run the heuristic functions to identify it. This is exactly what Wireshark does: it first tries fast port-based classification, and if that fails, it calls each protocol’s heuristic dissector to detect the packet more accurately.

@sorooshm78 I think this aligns with @Dimi1010 's suggestion above ☝️ The main issue I see with this approach is that all of the non-classified protocols (meaning protocols PcapPlusPlus doesn't yet parse) will fall into this bucket and will be checked for SIP matching (and in the future, maybe more protocols). 

Maybe we can combine this with option (2) - we can have add 2 static `SipLayer::parse()` methods with a signature like this:
```cpp
// Parse with checking the port
static SipLayer* parseSipLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet, uint16_t srcPort, uint16_t dstPort);

// Parse without checking the port
static SipLayer* parseSipLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet);
```

The first method will do roughly what the current parsing logic does: check the ports first, then check if it's a request or a response. It will be called instead of the current SIP parsing.

The second method will do what option (2) suggests: check the first 3 characters: if it looks like a SIP response - continue checking the version, status code, etc. If it looks like a request - continue checking the SIP method. This method will be called last - after all the port checks.

Please let me know what you think.

seladb commented:
@sorooshm78 @Dimi1010 my suggestion is very similar to what both of you suggest:
```cpp
// Parse with checking the port
static SipLayer* parseSipLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet, uint16_t srcPort, uint16_t dstPort);

// Parse without checking the port
static SipLayer* parseSipLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet);
```

The first overload is the "parse SIP by port" which will be called in the same place it's called today.

The second overload is the one who does "heuristic parsing" and will be called at the end, after the port search is exhausted. 

I agree we can think of a more generic approach which will have heuristics to parse more protocols, but for now maybe we can start with SIP only and see how it goes.


> For example, in my initial SIP heuristic I was taking the first line, splitting it into three tokens, and then parsing each one, which is more expensive. A faster approach is to just look at the first 3 characters of the payload:
> 
>     * For a response, the first 3 characters will be `SIP`.
> 
>     * For a request, the first token is the SIP method, and we can check only the first 3 characters (e.g. `INV`, `REG`, `BYE`, etc.) against the known SIP methods.
> 
> 
> This keeps the heuristic very cheap, but still doesn’t rely on the port.

@sorooshm78 I agree that could be a good heuristic that can minimize the performance impact.

sorooshm78 commented:
@seladb
Sounds good, I’ll update the PR with these changes soon.

Dimi1010 commented on Packet++/src/SipLayer.cpp:
```diff
@@ -103,66 +120,48 @@ namespace pcpp
-
-		return false;
+		return SipLayer::SipParseResult::Unknown;
 	}
```
Honestly. I am not sure this is better solution... It might be faster in theory, but it is also more leaky as it potentially allows malformed SIP packets? I am unsure if the SIP Layer is prepared to handle these? @seladb do you know?

IMO, changing the old `dissectSipHeuristic` to return the new `SipLayer::SipParseResult` to remove the need for a second parse to determine the precise SIP message type would have been enough, and it would have maintained the SIP packet integrity strictness.

sorooshm78 replied:
>Honestly. I am not sure this is a better solution than the previous dissector implementation... It might be faster in theory, but it is also more leaky as it potentially allows malformed SIP packets? 

Given the current code, the point you mentioned doesn’t fully hold. It’s true that `dissectSipHeuristic` cannot make a definitive determination, but the constructor of the `SipRequestFirstLine` and `SipResponseFirstLine` class is resilient against malicious SIP packets. Of course, I don’t deny that a SIP packet could still be malformed.

 https://github.com/seladb/PcapPlusPlus/blob/dev/Packet%2B%2B/src/SipLayer.cpp#L108-L141

sorooshm78 commented on Packet++/src/SipLayer.cpp:
```diff
@@ -103,66 +120,48 @@ namespace pcpp
+		if (sipParseResult == SipLayer::SipParseResult::Response)
+			return new SipRequestLayer(data, dataLen, prevLayer, packet);
+		return nullptr;
 	}
```
> @seladb Did I understand your point correctly? Is this the kind of code you mean?
If the answer is yes, in the function parseSipLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet, uint16_t srcPort, uint16_t dstPort), we are not using dstPort and srcPort unless we also perform isSipPort inside parseSipLayer, which would change the structure.

I also mentioned this.

seladb commented:
> @seladb Did I understand your point correctly? Is this the kind of code you mean?
> If the answer is yes, in the function `parseSipLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet, uint16_t srcPort, uint16_t dstPort)`, we are not using `dstPort` and `srcPort` unless we also perform `isSipPort` inside `parseSipLayer`, which would change the structure.

@sorooshm78 I'd do something slightly different:

```cpp
static SipLayer* tryParseSipLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet, uint16_t srcPort, uint16_t dstPort)
{
    if !(SipLayer::isSipPort(portDst) || SipLayer::isSipPort(portSrc))
    {
        return nullptr;
    }

    if (SipRequestFirstLine::parseMethod((char*)udpData, udpDataLen) != SipRequestLayer::SipMethodUnknown)
    {
        return new SipRequestLayer(udpData, udpDataLen, this, m_Packet);
    }

    if (SipResponseFirstLine::parseStatusCode((char*)udpData, udpDataLen) != SipResponseLayer::SipStatusCodeUnknown && SipResponseFirstLine::parseVersion((char*)udpData, udpDataLen) != "")
    {
        return new SipResponseLayer(udpData, udpDataLen, this, m_Packet);
    }
    
    return nullptr;
}

static SipLayer* parseSipLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet)
{
    // check the first 3 characters, guess if it's a request or response

    // if it's none of them -> return nullptr

    // if it looks like a request -> try to parse as much as possible from the first line using the existing methods (or refactor them a bit)
    // if it's indeed a request -> return new SipRequestLayer(data, dataLen, prevLayer, packet)

    // if it looks like a response ->try to parse as much as possible from the first line using the existing methods (or refactor them a bit)
    // if it's indeed a response -> return new SipRequestLayer(data, dataLen, prevLayer, packet)

    // return nullptr
}

void UdpLayer::parseNextLayer()
{
    uint16_t portDst = getDstPort();
    uint16_t portSrc = getSrcPort();

    /* Other if (skipPortCheck || TLayer::isTPort(portSrc) || TLayer::isTPort(portDst)) */

    else if (SipLayer::isSipPort(portDst) || SipLayer::isSipPort(portSrc))
        m_NextLayer = parseSipLayer(udpData, udpDataLen, this, m_Packet, portSrc, portDst);

    // Remove this code 
    // else
    //     m_NextLayer = new PayloadLayer(udpData, udpDataLen, this, m_Packet);
  
    // If a valid layer was found, return immediately
    if (m_NextLayer)
        return;

    // Here, heuristics for all protocols should be invoked to determine the correct layer
    m_NextLayer = parseSipLayer(udpData, udpDataLen, this, m_Packet);

    if (!m_NextLayer)
        m_NextLayer = new PayloadLayer(udpData, udpDataLen, this, m_Packet);
}
```

sorooshm78 commented:
@seladb, @Dimi1010 I’ve implemented the changes based on your suggestions; please review the code and let me know if any further adjustments are needed.

Dimi1010 reviewed (commented):
Looks good.

Personally I am still unsure why we need the second `parseSipLayer` function with `srcPort` and `dstPort` data input. With how the 2 methods are implemented currently, it looks like the main difference is that the port version just skips the initial `detectSipMessageType`. Since both versions now do a full parse before constructing the concrete `SipLayer` the skip could instead be implemented as a bool flag into the non-port version?

The `if (!(SipLayer::isSipPort(srcPort) || SipLayer::isSipPort(dstPort)))` branch will never actually be hit under current usages, as that is already guarded in the `UdpLayer`. So that is a bit of a non-factor difference (and minor overhead) IMO.

seladb commented on Packet++/header/SipLayer.h:
```diff
@@ -69,6 +69,17 @@ namespace pcpp
+			Request = 1,
+			/// The message is a SIP response
+			Response = 2,
+		};
```
This can be a private enum, no?

If we move it to be private, doxygen documentation is not required

seladb commented on Packet++/src/SipLayer.cpp:
```diff
@@ -103,6 +120,85 @@ namespace pcpp
+
+		if (SipResponseFirstLine::parseStatusCode(reinterpret_cast<char*>(data), dataLen) !=
+		        SipResponseLayer::SipStatusCodeUnknown &&
+		    SipResponseFirstLine::parseVersion(reinterpret_cast<char*>(data), dataLen) != "")
```
nit: you can check:
```
!SipResponseFirstLine::parseVersion(reinterpret_cast<char*>(data), dataLen).empty()
```

seladb commented on Packet++/src/SipLayer.cpp:
```diff
@@ -103,6 +120,85 @@ namespace pcpp
+		return nullptr;
+	}
+
+	SipLayer* SipLayer::parseSipLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet)
```
ditto: you can use `empty()` in this method instead of `!= ""`

Dimi1010 commented on Packet++/src/SipLayer.cpp:
```diff
@@ -103,6 +120,85 @@ namespace pcpp
+			if (SipRequestFirstLine::parseMethod(reinterpret_cast<char*>(data), dataLen) !=
+			        SipRequestLayer::SipMethodUnknown &&
+			    SipRequestFirstLine::parseUri(reinterpret_cast<char*>(data), dataLen) != "" &&
+			    SipRequestFirstLine::parseVersion(reinterpret_cast<char*>(data), dataLen) != "")
```
Hmm, if we want to extract the uri / version from raw data statically, I think a combined function would be better. Possibly one that returns a `struct PlaceholderName { std::string uri; std::string version }`. That way we only do a single parse run on the line. If partial parse is needed we can take an enum flag on which fields to populate, but that is an optional refinement. Further processing on validation can be done on the returned struct. For example: return empty fields on invalid data.

Having a separate value only struct representation of the entire first line would also allow a bulk replacement via the `Sip***FirstLine` proxy objects, potentially optimizing the inefficiencies that many of the setters have by shortening / extending the layer in every setter call.

sorooshm78 replied:
@seladb  
I agree with @Dimi1010 ’s suggestion. If you also agree, we can merge these two functions into one that parses the entire first line and returns a struct like:

```cpp
struct SipFirstLineData {
    std::string uri;
    std::string version;
};
```

This way we only scan the line once, which is more efficient, and we still keep the ability to extract URI and version statically.

sorooshm78 commented on Packet++/header/SipLayer.h:
@seladb
I’ve added `SipDetectionByContentOnNonStandardPort` in `SipSdpTests.cpp` to test SIP detection by content on non-standard ports.
Let me know if you’d like any adjustments.

Dimi1010 commented on Packet++/header/SipLayer.h:
```diff
@@ -498,6 +534,16 @@ namespace pcpp
+			std::string method;   ///< The SIP method (e.g., INVITE, REGISTER, BYE)
+			std::string uri;      ///< The Request-URI destinati
+			std::string version;  ///< The SIP protocol version (e.g., SIP/2.0)
+			bool isValid;         ///< True if parsing succeeded, false otherwise
```
Hmm, I think extracting the isValid outside the data struct would be better semantics and returning something like std::pair<bool, SipFirstLineData> (or std::optional<SipFirstLineData> if we had C++17) from parseFirstLine(const char* data, size_t dataLen).

Since the POD struct offers no validation on data member mutations, the isValid flag can/will quickly become stale / meaningless after the parse operation, and could no longer be relied on.

sorooshm78 replied:
You're absolutely right - I've updated to `std::pair<bool, SipFirstLineData>` to separate the validation result from the data.

Dimi1010 commented on Packet++/src/SipLayer.cpp:
```diff
@@ -208,6 +296,85 @@ namespace pcpp
+		}
+
+		// All validations passed
+		result.method = methodStr;
```
nit: Use move assignment to avoid a copy.
```suggestion
		result.method = std::move(methodStr);
```

sorooshm78 replied:
Thanks for the correction and assembly proof! You're right - `std::move` is needed here. The assembly clearly shows copy vs. move assignment. My mistake!

Dimi1010 commented on Packet++/header/SipLayer.h:
```diff
@@ -498,6 +534,15 @@ namespace pcpp
+		struct SipFirstLineData
+		{
+			std::string method;   ///< The SIP method (e.g., INVITE, REGISTER, BYE)
+			std::string uri;      ///< The Request-URI destinati
```
nit: Typo
```suggestion
			std::string uri;      ///< The Request-URI destination
```

Dimi1010 reviewed (approved):
LGTM

seladb commented on Packet++/src/UdpLayer.cpp:
```diff
@@ -109,16 +109,7 @@ namespace pcpp
-			else
-				m_NextLayer = new PayloadLayer(udpData, udpDataLen, this, getAttachedPacket());
-		}
+			m_NextLayer = SipLayer::parseSipLayer(udpData, udpDataLen, this, getAttachedPacket(), portSrc, portDst);
```
If `SipLayer::parseSipLayer()` returns `nullptr` we can call `constructNextLayer<PayloadLayer>(udpData, udpDataLen, getAttachedPacket());`

sorooshm78 replied:
I wrote it this way to maintain this particular code style.
```
if (!m_NextLayer)
    m_NextLayer = constructNextLayer<PayloadLayer>(udpData, udpDataLen, getAttachedPacket());
```

Dimi1010 replied:
> I wrote it this way to maintain this particular code style.
> 
> ```
> if (!m_NextLayer)
>     m_NextLayer = constructNextLayer<PayloadLayer>(udpData, udpDataLen, getAttachedPacket());
> ```

@sorooshm78 This is wrong. `constructNextLayer` automatically sets the `m_NextLayer` variable. No need to set it again.

seladb commented on Packet++/src/UdpLayer.cpp:
```diff
@@ -152,7 +143,15 @@ namespace pcpp
+		// Here, heuristics for all protocols should be invoked to determine the correct layer
+		m_NextLayer = SipLayer::parseSipLayer(udpData, udpDataLen, this, getAttachedPacket());
+
+		if (!m_NextLayer)
```
nit: we usually surround even one line with curly braces `{}`:
```cpp
if  (m_NextLayer)
{
	return;
}

...

if (!m_NextLayer)
{
	m_NextLayer = new PayloadLayer(udpData, udpDataLen, this, getAttachedPacket());
}
```

sorooshm78 commented on Packet++/src/SipLayer.cpp:
```diff
@@ -208,6 +296,85 @@ namespace pcpp
+		// Validate method exists in SipMethodStringToEnum
+		std::string methodStr{ data, firstSpaceIndex };
+		auto methodIt = SipMethodStringToEnum.find(methodStr);
+		if (methodIt == SipMethodStringToEnum.end())
```
Fixed

sorooshm78 commented on Packet++/src/SipLayer.cpp:
```diff
@@ -103,6 +120,77 @@ namespace pcpp
+		if (it != SipMethodShortMap.end())
+		{
+			return SipLayer::SipParseResult::Request;
+		}
```
Fixed

seladb commented on Packet++/src/SipLayer.cpp:
```diff
@@ -103,6 +120,76 @@ namespace pcpp
 		}
 	}
 
+	SipLayer::SipParseResult SipLayer::detectSipMessageType(const uint8_t* data, size_t dataLen)
```
Since this method will be called for each UDP packet that wasn't classified by another protocol, I asked Claude AI if we can optimize it even more. I suggested the following:

1. Avoid the `std::string` allocation - This is the biggest bottleneck. Creating a temporary string allocates memory on every call.
2. Use integer comparison instead of byte-by-byte checks - Load and compare 4 bytes at once.
3. Replace map lookup with a switch or perfect hash - Map lookups are relatively slow for small sets.

Here is an alternative implementation it suggested:
```cpp
SipLayer::SipParseResult SipLayer::detectSipMessageType(const uint8_t* data, size_t dataLen)
{
    if (!data || dataLen < 3)
    {
        return SipLayer::SipParseResult::Unknown;
    }

    // Load first 4 bytes as uint32_t (assumes little-endian, adjust if needed)
    uint32_t first4;
    memcpy(&first4, data, sizeof(uint32_t));
    
    // Check for "SIP/" (0x2F504953 on little-endian)
    if ((first4 & 0x00FFFFFF) == 0x00504953) // Mask to compare only 3 bytes
    {
        return SipLayer::SipParseResult::Response;
    }

    // Check common SIP methods by their first 3 bytes
    uint32_t first3 = first4 & 0x00FFFFFF;
    switch (first3)
    {
        case 0x00564E49: // "INV" (INVITE)
        case 0x00544547: // "GET"
        case 0x00544550: // "PET" (OPTIONS - need to verify 4th byte)
        case 0x00474552: // "REG" (REGISTER)
        case 0x00435553: // "SUC" (SUBSCRIBE - need to verify)
        case 0x00595442: // "BTY" (BYE)
        case 0x00434143: // "CAC" (CANCEL)
        case 0x004B4341: // "ACK"
        // Add other methods as needed
            return SipLayer::SipParseResult::Request;
    }

    return SipLayer::SipParseResult::Unknown;
}
```

If we implement it we can probably remove `SipMethodShortMap`

sorooshm78 replied:
@seladb I understand your concern about performance here, especially since this code runs on every UDP packet that wasn’t classified by another protocol, so squeezing out overhead makes sense.

That said, I agree with @Dimi1010 ’s points regarding the endianness of integer literals. The fully “magic-number” approach becomes harder to reason about and less portable, which is a real maintenance cost.

The final `pack4`-based solution feels like a very good compromise to me: it avoids `std::string` allocations, removes the map lookup, keeps the comparisons integer-based, and stays readable and endian-safe. From my perspective, that last version hits the right balance between performance, clarity, and portability.

seladb replied:
Thank you @Dimi1010 ! Your solution seems great 💯

seladb reviewed (commented):
@sorooshm78 please find last few comments, this PR is almost ready to be merged 👍

sorooshm78 commented on Tests/Packet++Test/Tests/SipSdpTests.cpp:
```diff
@@ -166,6 +166,41 @@ PTF_TEST_CASE(SipRequestLayerParsingTest)
+	PTF_ASSERT_TRUE(sipReqNonStandardPort.isPacketOfType(pcpp::SIP));
+	PTF_ASSERT_TRUE(sipReqNonStandardPort.isPacketOfType(pcpp::SIPRequest));
+
+	pcpp::SipRequestLayer* sipReqLayer = sipReqNonStandardPort.getLayerOfType<pcpp::SipRequestLayer>();
```
Fixed

sorooshm78 commented on Tests/Packet++Test/Tests/SipSdpTests.cpp:
```diff
@@ -166,6 +166,41 @@ PTF_TEST_CASE(SipRequestLayerParsingTest)
+	auto rawPacket1 = createPacketFromHexResource("PacketExamples/sip_non_default_port1.dat");
+	pcpp::Packet sipReqNonStandardPort(rawPacket1.get());
+
+	PTF_ASSERT_TRUE(sipReqNonStandardPort.isPacketOfType(pcpp::SIP));
```
Fixed

sorooshm78 commented on Packet++/src/UdpLayer.cpp:
```diff
@@ -110,14 +110,9 @@ namespace pcpp
-			else
-				m_NextLayer = new PayloadLayer(udpData, udpDataLen, this, getAttachedPacket());
+			m_NextLayer = SipLayer::parseSipLayer(udpData, udpDataLen, this, getAttachedPacket(), portSrc, portDst);
+			if (!m_NextLayer)
```
Fixed

Dimi1010 commented on Packet++/src/SipLayer.cpp:
```diff
@@ -103,6 +103,92 @@ namespace pcpp
+		{
+		case "INVI"_packed4:  // INVITE
+		case "ACK "_packed4:  // ACK
+		case "BYE "_packed4:  // BYE
```
A question. Is ACK and BYE always followed by a space? If so, then this is fine.

If not, this might have issues if `dataLen == 3`, where the 4th byte will be '\0' instead of ' ' on the `key`.

seladb commented on Packet++/header/SipLayer.h:
```diff
@@ -8,6 +8,26 @@
+	{
+		return pack4(str, len);
+	}
+
```
@sorooshm78 I agree with @Dimi1010 that it should move to the `.cpp` file, I missed that in my review

Dimi1010 commented:
@sorooshm78 PR has been merged. Thank you for working on this. 🙂

sorooshm78 commented:
@seladb Thanks a lot! No worries at all — I really appreciate the review and the feedback, and I’d be happy to contribute more to this project in the future.
